In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
#getting data in jupyter
path = r'C:\Users\mande\OneDrive\Documents\Instacart Basket folder 20th April 2023'

In [3]:
# Import CSV file using path
df_prods = pd.read_csv(os.path.join(path, 'Data', 'original data', 'products.csv'), index_col = False)

In [4]:
#importing  from orders_wrangled.csv  prepared data set
df_ord = pd.read_csv(os.path.join(path, 'Data', 'prepared data', 'orders_wrangled.csv'), index_col = False)

checking mixed data type through a new data frame

In [5]:
df_test = pd.DataFrame()

In [6]:
df_test['Mix'] =['a', 'b', 1, True]

In [7]:
df_test.head()

Mix
0     a
1     b
2     1
3  True

checking for mixed data type in the above dataset

In [8]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

Mix


In [9]:
#findig missing where are missing value
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
df_prods.dropna(subset = ['product_name'], inplace = False)

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

checking only mean value and median value in below two cells of data

In [13]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [14]:
df_prods.mean()

C:\Users\mande\AppData\Local\Temp\ipykernel_22412\730692063.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods.mean()


product_id       24844.345139
aisle_id            67.770249
department_id       11.728433
prices               9.994136
dtype: float64

In [15]:
 value = df_prods.median()

C:\Users\mande\AppData\Local\Temp\ipykernel_22412\1740243667.py:1: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  value = df_prods.median()


Although we got both median and mean of df_prods but our missing values are from product_name which can be filled with mean or median as its a string and mean and median are a integiner or float values. 


In [16]:
df_prods.shape

(49693, 5)

command to drop only the NaNs from a particular column,

In [17]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

checking how many row and column we have after getting rid of nan values from product_name

In [18]:
df_prods_clean.shape

(49677, 5)

Checking for duplicate in our new data made in above cell

In [19]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

addressing duplicate rows

In [21]:
df_prods_clean_no_dups =df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

Below I am changing one of df_ord data's coulmn name. somehow it havn't been updated from previous task.

In [23]:
df_ord.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [24]:
#using describe function
df_ord.describe()


Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.029782e+05  1.029782e+05  1.715486e+01   
std    9.875817e+05  5.953372e+04  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  5.139400e+04  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.026890e+05  1.026890e+05  1.100000e+01   
75%    2.565812e+06  1.543850e+05  1.543850e+05  2.300000e+01   
max    3.421082e+06  2.062090e+05  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_last_order  
count       3.421083e+06       3.421083e+06           3.214874e+06  
mean        2.776219e+00       1.345202e+01           1.111484e+01  
std         2.046829e+00       4.226088e+00           9.206737e+00  
min         0.000000e+00       0.000000e+00           0.000000e+00  
25%         1.000000e+00       1.000000e+01           4.000000e+00  
50%         3.000000e+00       1.300000e+01           7.000000e+00  
75%         5.000000e+00       1.600000e+01           1.500000e+01  
max         6.000000e+00       2.300000e+01           3.000000e+01

checking our row and column in df_ord dataframe

In [33]:
df_ord.shape

(3421083, 8)

#checking mix type data in df_ord


In [25]:
for col in df_ord.columns.tolist():
  weird = (df_ord[[col]].applymap(type) != df_ord[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ord[weird]) > 0:
    print (col,' mixed')
  else: print (col, ' consistent')
 

Unnamed: 0  consistent
order_id  consistent
user_id  consistent
eval_set  consistent
order_number  consistent
order_day_of_week  consistent
order_hour_of_day  consistent
days_since_last_order  consistent


In [26]:
df_ord.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
eval_set                      0
order_number                  0
order_day_of_week             0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

In [30]:
df_ord_nan = df_ord[df_ord['days_since_last_order'].isnull() == True]

In [31]:
df_ord_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0         1        1    prior             1   
11               11         2        2    prior             1   
26               26         3        3    prior             1   
39               39         4        4    prior             1   
45               45         5        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    206205   206205    prior             1   
3420934     3420934    206206   206206    prior             1   
3421002     3421002    206207   206207    prior             1   
3421019     3421019    206208   206208    prior             1   
3421069     3421069    206209   206209    prior             1   

         order_day_of_week  order_hour_of_day  days_since_last_order  
0                        2                  8                    NaN  
11                       2                 11                    NaN  
26                       1                 14                    NaN  
39                       6                 11                    NaN  
45                       3                 12                    NaN  
...                    ...                ...                    ...  
3420930                  4                 12                    NaN  
3420934                  3                 18                    NaN  
3421002                  6                 19                    NaN  
3421019                  1                 15                    NaN  
3421069                  3                 11                    NaN  

[206209 rows x 8 columns]

above data is showing 206209 missing values. All these missing values from the column which is showing orders since last order column. if we see order number which is 1 so before that order there wasn't any order from the users. In the case we can say that user must joined and days since last order column is counting the days even before their first order.

Along witht that column is showing a huge number to replace. Mostly we should replace 5% or less if we find missing value. We have total
3421083 rows and missing values are 206209. so I am not going to change missing value as it still telling us that there was no order perior since last day.


In [35]:
#finding duplicates in df_ord
df_ord_dup = df_ord[df_ord.duplicated()]

In [36]:
df_ord_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [37]:
df_ord_dup.shape

(0, 8)

we don't have any duplicate in df_ord data set as it is showing 0 for rows and showing 8 coulms.

In [38]:
df_ord.to_csv(os.path.join(path, 'Data','Prepared data', '4.5 Data Consistency Checks.csv'))

In [40]:
df_prods.to_csv(os.path.join(path, 'Data','Prepared data', '4.5 Data Consistency Checks.csv'))